src/gensxg.c

```diff
@@ -0,0 +1,444 @@
+    valid = false;
+  }
+  if (opt->mi_record_size == 0) {
+    fprintf(stderr, "error: -miRecordSize must be greater than 0.\n");
+    valid = false;
+  }
+  if (is_empty(opt->url)) {
+    fprintf(stderr, "error: -url must not be empty.\n");
+    valid = false;
+  }
+  if (is_empty(opt->cert_url)) {
+    fprintf(stderr, "error: -certUrl must not be empty.\n");
+    valid = false;
+  }
+  if (is_empty(opt->validity_url)) {
+    fprintf(stderr, "error: -validityUrl must not be empty.\n");
+    valid = false;
+  }
+  if (is_empty(opt->certificate)) {
+    if (is_empty(opt->public_key)) {
+      fprintf(stderr,
+              "error: -certificate or -publicKey must be specified.\n");
+      valid = false;
+    } else if (access(opt->public_key, F_OK)) {
+      fprintf(stderr,
+              "error: Cannot access publicKey: %s\n",
+              opt->public_key);
+      valid = false;
+    }
+  } else {
+    if (opt->public_key != NULL) {
+      fprintf(stderr,
+              "error: -certificate and -publicKey cannot be "
+              "specified at the same time.\n");
+      valid = false;
+    } else if (access(opt->certificate, F_OK)) {
+      fprintf(stderr,
+              "error: Cannot access certificate: %s\n",
+              opt->certificate);
+      valid = false;
+    }
+  }
+  if (is_empty(opt->private_key)) {
+    fprintf(stderr, "error: -privateKey must be specified.\n");
+    valid = false;
+  } else if (access(opt->private_key, F_OK)) {
+    fprintf(stderr,
+            "error: Cannot access privateKey: %s\n",
+            opt->private_key);
+    valid = false;
+  }
+  if (opt->duration <= 0) {
+    fprintf(stderr, "error: SXG lifespan must be a positive number.\n");
+  }
+  return valid;
+}
+
+static void load_signer(const Options* opt, sxg_signer_list_t* signers) {
+  uint64_t date = (uint64_t)opt->date;
+  uint64_t expire = (uint64_t)opt->date + (uint64_t)opt->duration;
+
+  if (opt->certificate != NULL) {
+    X509* cert = load_x509_cert(opt->certificate);
+    EVP_PKEY* pri_key = load_private_key(opt->private_key);
+    if (!sxg_add_ecdsa_signer(opt->url, date, expire,
+                              opt->validity_url, pri_key,
+                              cert, opt->cert_url, signers)) {
+      fprintf(stderr, "Failed to generate SXG signer with certificate\n");
+      exit(EXIT_FAILURE);
+    }
+    X509_free(cert);
+    EVP_PKEY_free(pri_key);
+  } else {
+    EVP_PKEY* pub_key = load_ed25519_pubkey(opt->public_key);
+    EVP_PKEY* pri_key = load_private_key(opt->private_key);
+    if (!sxg_add_ed25519_signer(opt->url, date, expire,
+                                opt->validity_url, pri_key,
+                                pub_key, signers)) {
+      fprintf(stderr, "Failed to generate SXG signer with Ed25519 key\n");
+      exit(EXIT_FAILURE);
+    }
+    EVP_PKEY_free(pub_key);
+    EVP_PKEY_free(pri_key);
+  }
+}
+
+static void load_contents(const Options* opt, sxg_buffer_t* buf) {
```

Comment:
`static void load_content(const char* filepath, sxg_buffer_t* buf)`?

Remove the plural ("s"). Take the filename rather than the Options.

Reply:
Done.